docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+
+### 6. Report the diagnostic only on parts of code that would be changed by a code fixer.
+ This is a better communication to the end-user when running in Visual Studio and it makes it easier to write the Code Fixer.
+
+### 7. Do not use async methods in the synchronous context of an Analyzer. 
+If the only methods available are async, you should find another way to implement your analyzer to prevent runtime issues.
+
+**Do not**
+* Use `.Result` or `Wait()` in a Roslyn Analyzer. Forcing asynchronous code the behave synchronously can result in timing issues that are hard to debug and thread starvation.
+
+## Best Practices for Roslyn Code Fixers
+
+### 1. Code fixers handle well-known scenarios
+You may need to do a few final checks before registering your code fixer by calling `RegisterCodeFix` but you should lean into doing as many checks as possible in the analyzer.
+
+You can also accelerate your code fixer productivity by using the additionalLocations overload when reporting a diagnostic so that you can find multiple interesting locations in a document when working from a single diagnostic.
+
+### 2. Use the `SyntaxGenerator` to create language agnostic code fixers
+In many scenarios, your code fixer can apply to Csharp and VB. The SyntaxGenerator  is a language agnostic factory for creating syntax nodes.
+ 
+The trees generated by this API will try to respect user preferences when possible. For example, generating MemberAccessExpression(SyntaxNode, String) will be done in a way such that `this.` or `Me.` will be simplified according to user preference if any `ReduceAsync(Document, OptionSet, CancellationToken)` overload is called.
+
+### 3. Beware of Trivia
+When replacing a SyntaxNode consider whether or not the node you're replacing had trivia. Trivia, the spacing we use to make code more readable, can vary from team to team. Not everyone puts a space after an assignment operator and if there are two spaces, or a tab, after the assignment operator then the code fixer should respect the file's original formatting.
+
+**Do**
+* Use the `WithLeadingTrivia` and `WithTrailingTrivia` extension methods to preserve trivia when creating new nodes.
+
+### 4. Apply Async programming best practices
+Code fixers are asynchronous. Follow best practice guidance for asynchronous programming to prevent race conditions and performance issues.
+
+Learn more about Async best practices by reading [Async/Await - Best Practices in Asynchronous Programming](https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming) by Stephen Cleary
+
+**Do**
+* Leverage the CancellationToken to interrupt processing when cancellation has been requested.
+
+**Do not**
+* Do not force async code to run synchronously. Avoid calling `.Result` and prefer to use await as needed when working in the code fixer construct.
+
+### 5. When using CodeAction.Create prefer Document over Solution
```

Comment:
Is this a best practice? What about limiting the smallest scope required?

Reply:
That's the intent of this one, I'll reword the title